Add endpoint for chat history

mzgajner opened the pull request:
Nadaljevanje API refactoringa - dodal sem endpoint za listanje chat messagev (prej ni obstajal, ker se je history poslal kar prek socketa on connect).

Obenem sem dodal še en lib za validacijo request parametrov in začel malo improvati errorje v responseih. Kmalu namečem tut na ostale endpointe.

Če bo kdo reviewal, priporočam po commitih, ker je na koncu en še kar velik formatting change.

zigomir commented on backend/lib/kjer_si_web/views/error_view.ex:
```diff
@@ -13,4 +13,17 @@ defmodule KjerSiWeb.ErrorView do
+    %{
+      errors: %{
+        detail: Plug.Conn.Status.reason_phrase(422),
+        fields: translate_errors(changeset)
```
A to potem pove kater field je napacn oz manjka?

S testov nism razbral kaj bi naj vrnil. Maybe dodaj v kak test case z `errors.fields`.

mzgajner replied:
Ja, tocno to. Great point, seveda bi moral potestirat za to, fixed :+1:

zigomir commented on backend/lib/kjer_si_web/router.ex:
```diff
@@ -24,12 +24,15 @@ defmodule KjerSiWeb.Router do
     post "/users", UserController, :create
     get "/categories", RoomController, :categories
-    post "/rooms", RoomController, :create
-    post "/map/rooms", MapController, :get_rooms_in_radius
```
👍 coolio – less controlling, more action!

zigomir commented on backend/lib/kjer_si_web/router.ex:
```diff
@@ -24,12 +24,15 @@ defmodule KjerSiWeb.Router do
       get "/users/self", UserController, :self
+
+      resources "/rooms", RoomController, only: [:index, :create] do
+        resources "/messages", MessageController, only: [:index]
```
Mogoce tu, kjer imamo samo`:index` lahko kar uporabis `get` namesto `resources`?

mzgajner commented on backend/lib/kjer_si/messages.ex:
```diff
@@ -17,8 +17,24 @@ defmodule KjerSi.Messages do
+        query
+      end
+
+    Repo.all(query)
```
To z bolj functional approachem je super komentar :heart:, sploh še nisem zares internaliziral tega, da lahko definiram istoimenske funkcije z različnim arityjem al pa z malo drugačnimi patterni v argumentih.

V tem primeru drugače ugotavljam, da sem itak brezveze kompliciral, ker sta oba parametra mandatory v kontrolerju, tako da zaenkrat bom samo ven pobral to dinamično generiranje.

mzgajner commented on backend/test/kjer_si/messages_test.exs:
```diff
@@ -3,42 +3,61 @@ defmodule KjerSi.MessagesTest do
+
+    test "list_messages accepts 'before' as first argument", %{user_id: user_id, room_id: room_id} do
+      _message1 = message_fixture(user_id, room_id)
+      _message2 = message_fixture(user_id, room_id)
```
:man_facepalming:

mzgajner commented:
OK, addressed the comments (zadnjih 5 commitov), rebasal na latest master (konflikti) in našel ter fixnil še en serious bug - v `list_messages` sploh nisem filtriral po sobi, samo slučajno je imel ves seed in test data mesiđe sam za eno sobo :blush:. Dodal tut teste za to.

Dve vprašanji mam zate:

Ko zgeneriraš neko novo shemo s Phoenixom, dobiš zraven tut vse osnovne CRUD metode. V štartu smo to puščali not _za vsak slučaj_, pa ker je blo fajn met par lepih primerov za zgled. Zdaj se mi zdi, da počasi štekamo kaj bo potrebno in kaj ne. A je kakorkoli sporno, da to sproti mečem ven (npr. [predzadnji commit](https://github.com/danesjenovdan/kjer.si/pull/14/commits/96ed64497ca284fe2d47c8c77c67bdbffec7ee9a))?

Ene `Ecto.Repo` funkcije (npr. `insert`) vračajo `{:ok, result}/{:error, error}`, druge (npr. `get`) pa ne. Ta prvi tip responsea se mi zdi ful praktičen za clean generic error handling s pomočjo `with` v kontrolerjih. A je legit pol mal powrappat stvari in na roke [vrnit tak tuple](https://github.com/danesjenovdan/kjer.si/pull/14/commits/2a676bc925bbe0318ba4403072374415a3e01dd3#diff-913dae3bee629cc4b275bd9e75182f0aR153-R158)?

mzgajner commented on backend/lib/kjer_si_web/controllers/message_controller.ex:
```diff
@@ -0,0 +1,28 @@
+
+    if changeset.valid? do
+      messages =
+        Messages.list_messages(params["before"], params["limit"])
```
En response višje :point_up:

zigomir commented:
Huh, nice find. Sm vidu nested route pa kr assumal da je all good :)

1. men je kul da se spuca sprot. Brez veze bit prevec cistunski (pa vse dat v svoj PR), ker pol se sam prelaga stvari na kasneje :)

2. fuf. men zgleda legit. me pa zdej predvsem zanima kako to da je temu tako. Sem zdej najdu tole: https://elixirforum.com/t/why-does-repo-get-and-repo-get-by-return-the-resource-or-nil-instead-of-an-ok-or-error-tuple/23089 – morm se se prebit cez.

drugac pa ČMZV

mzgajner commented:
<3 tenks za ta forum post! Sem se včeraj pod večer prebil čez, pa ni blo ničesar zaskrbljujočega. Diskusija je predvsem okrog tega, zakaj nima sam Ecto konsistentnega patterna (par res dobrih insightov, ta peerreynders deluje kr car), ampak je dosti ljudi omenilo, da ima podoben usecase kot mi, da so si napisali wrapperje and everyone seems to think it's fine.
